File: client/session_manager.py
"""Ties crypto (X3DH + Double Ratchet) to local storage and the network.

This is the layer the GUI talks to. Every ratchet-encrypted plaintext is a
small JSON "envelope" (see _build_envelope) rather than a raw string, which
is what makes deletes, group messages and attachments possible on top of the
same pairwise Double Ratchet sessions built for 1:1 chat. Group conversations
are pure pairwise fan-out: one envelope (one id) gets encrypted separately to
every other member's existing/newly-established 1:1 session -- there is no
group-level cryptographic key.
"""
from __future__ import annotations

import base64
import json
import time
import uuid
from dataclasses import dataclass, field

from crypto import x3dh
from crypto.identity import KeyStore
from crypto.ratchet import RatchetMessage, RatchetState
from client.network import IncomingMessage, NetworkClient
from client.storage import (
    ConversationSummary,
    LocalStorage,
    StoredMessage,
    group_conversation_id,
    is_group_conversation,
)

ONE_TIME_PREKEY_TARGET = 20
ATTACHMENT_MAX_BYTES = 10 * 1024 * 1024
AVATAR_MAX_BYTES = 2 * 1024 * 1024


@dataclass
class IncomingEvent:
    kind: str  # "message" | "delete" | "group_invite" |
    # "group_member_added" | "group_member_removed" | "avatar_update"
    conversation_id: str
    sender: str
    timestamp: float
    message_id: str | None = None
    message_kind: str | None = None  # "text" | "attachment" | "system", when kind == "message"
    body: dict = field(default_factory=dict)


class SessionManager:
    def __init__(self, keystore: KeyStore, storage: LocalStorage, network: NetworkClient):
        self.keystore = keystore
        self.storage = storage
        self.network = network

    @property
    def username(self) -> str:
        return self.keystore.username

    async def publish_keys(self) -> None:
        """Upload identity/signed-prekey/one-time-prekeys; top up if running low."""
        if self.keystore.top_up_one_time_prekeys(ONE_TIME_PREKEY_TARGET):
            self.storage.save_keystore(self.keystore)
        await self.network.upload_keys(self.keystore)

    # -- recipient resolution --------------------------------------------------------

    def _recipients_for(self, conversation_id: str) -> list[str]:
        if is_group_conversation(conversation_id):
            group_id = conversation_id.split(":", 1)[1]
            return [m for m in self.storage.group_members(group_id) if m != self.username]
        return [conversation_id]

    # -- low-level: encrypt+send one envelope to one recipient -----------------------

    async def _send_envelope_to(self, recipient_username: str, envelope: dict) -> None:
        session = self.storage.load_session(recipient_username)
        x3dh_header_json = None

        if session is None:
            bundle = await self.network.fetch_bundle(recipient_username)
            if bundle is None:
                raise ValueError(f"no such user or no published keys: {recipient_username}")
            result, header = x3dh.initiate(self.keystore, bundle)
            session = RatchetState.init_as_initiator(
                shared_secret=result.shared_secret,
                associated_data=result.associated_data,
                remote_ratchet_pub=bundle.signed_prekey_pub,
            )
            x3dh_header_json = header.to_json()

        ratchet_message = session.encrypt(_envelope_bytes(envelope))
        payload = {"x3dh_header": x3dh_header_json, "ratchet_message": ratchet_message.to_json()}
        await self.network.send_message(recipient_username, payload)
        self.storage.save_session(recipient_username, session)

    async def _fan_out(self, conversation_id: str, envelope: dict) -> None:
        for recipient in self._recipients_for(conversation_id):
            await self._send_envelope_to(recipient, envelope)

    def _require_group_membership(self, conversation_id: str) -> None:
        if is_group_conversation(conversation_id):
            group_id = conversation_id.split(":", 1)[1]
            if self.username not in self.storage.group_members(group_id):
                raise ValueError("you are no longer a member of this group")

    # -- sending: text / attachment ---------------------------------------------------

    async def send_text(self, conversation_id: str, text: str) -> str:
        self._require_group_membership(conversation_id)
        message_id = str(uuid.uuid4())
        group_id = conversation_id.split(":", 1)[1] if is_group_conversation(conversation_id) else None
        envelope = {"id": message_id, "type": "text", "group_id": group_id, "body": {"content": text}}
        await self._fan_out(conversation_id, envelope)

        if not is_group_conversation(conversation_id):
            self.storage.add_contact(conversation_id)
        self.storage.add_message(
            conversation_id, self.username, "sent", "text", message_id, envelope["body"],
            timestamp=time.time(),
        )
        return message_id

    async def send_attachment(self, conversation_id: str, filename: str, mime: str, data: bytes) -> str:
        self._require_group_membership(conversation_id)
        if len(data) > ATTACHMENT_MAX_BYTES:
            raise ValueError(f"attachment too large ({len(data)} bytes, max {ATTACHMENT_MAX_BYTES})")
        message_id = str(uuid.uuid4())
        group_id = conversation_id.split(":", 1)[1] if is_group_conversation(conversation_id) else None
        body = {"filename": filename, "mime": mime, "size": len(data), "data_b64": base64.b64encode(data).decode()}
        envelope = {"id": message_id, "type": "attachment", "group_id": group_id, "body": body}
        await self._fan_out(conversation_id, envelope)

        if not is_group_conversation(conversation_id):
            self.storage.add_contact(conversation_id)
        self.storage.add_message(
            conversation_id, self.username, "sent", "attachment", message_id, body,
            timestamp=time.time(),
        )
        return message_id

    # -- delete-for-everyone -----------------------------------------------------------

    async def delete_message(self, conversation_id: str, message_id: str) -> None:
        self.storage.mark_deleted(conversation_id, message_id)
        group_id = conversation_id.split(":", 1)[1] if is_group_conversation(conversation_id) else None
        envelope = {"id": str(uuid.uuid4()), "type": "delete", "group_id": group_id,
                    "body": {"target_id": message_id}}
        await self._fan_out(conversation_id, envelope)

    # -- groups ---------------------------------------------------------------------------

    async def create_group(self, name: str, members: list[str]) -> str:
        group_id = str(uuid.uuid4())
        all_members = sorted(set(members) | {self.username})
        self.storage.create_group(group_id, name, all_members)
        conversation_id = group_conversation_id(group_id)

        envelope = {"id": str(uuid.uuid4()), "type": "group_invite", "group_id": group_id,
                    "body": {"group_id": group_id, "name": name, "members": all_members}}
        for member in all_members:
            if member != self.username:
                await self._send_envelope_to(member, envelope)

        self.storage.add_message(
            conversation_id, None, "sent", "system", str(uuid.uuid4()),
            {"content": "You created this group"}, timestamp=time.time(),
        )
        return group_id

    def groups(self) -> list[tuple[str, str]]:
        return self.storage.list_groups()

    async def add_group_member(self, group_id: str, username: str) -> None:
        conversation_id = group_conversation_id(group_id)
        current_members = self.storage.group_members(group_id)
        if username in current_members:
            return
        self.storage.add_group_member(group_id, username)
        new_roster = sorted(set(current_members) | {username})
        name = self.storage.group_name(group_id) or "Unnamed group"

        invite_envelope = {"id": str(uuid.uuid4()), "type": "group_invite", "group_id": group_id,
                            "body": {"group_id": group_id, "name": name, "members": new_roster}}
        await self._send_envelope_to(username, invite_envelope)

        added_envelope = {"id": str(uuid.uuid4()), "type": "group_member_added", "group_id": group_id,
                           "body": {"group_id": group_id, "added_username": username}}
        for member in current_members:
            if member != self.username:
                await self._send_envelope_to(member, added_envelope)

        self.storage.add_message(
            conversation_id, None, "sent", "system", str(uuid.uuid4()),
            {"content": f"You added {username}"}, timestamp=time.time(),
        )

    async def remove_group_member(self, group_id: str, username: str) -> None:
        conversation_id = group_conversation_id(group_id)
        old_roster = self.storage.group_members(group_id)
        if username not in old_roster:
            return
        self.storage.remove_group_member(group_id, username)

        removed_envelope = {"id": str(uuid.uuid4()), "type": "group_member_removed", "group_id": group_id,
                             "body": {"group_id": group_id, "removed_username": username}}
        for member in old_roster:
            if member != self.username:
                await self._send_envelope_to(member, removed_envelope)

        note = "You left this group" if username == self.username else f"You removed {username}"
        self.storage.add_message(
            conversation_id, None, "sent", "system", str(uuid.uuid4()),
            {"content": note}, timestamp=time.time(),
        )

    # -- profile pictures -------------------------------------------------------------------

    def _known_recipients(self) -> set[str]:
        recipients = set(self.storage.list_contacts())
        for group_id, _name in self.storage.list_groups():
            recipients.update(self.storage.group_members(group_id))
        recipients.discard(self.username)
        return recipients

    async def set_profile_picture(self, image_bytes: bytes, mime: str) -> None:
        if len(image_bytes) > AVATAR_MAX_BYTES:
            raise ValueError(f"image too large ({len(image_bytes)} bytes, max {AVATAR_MAX_BYTES})")
        self.storage.set_avatar(self.username, image_bytes)
        envelope = {"id": str(uuid.uuid4()), "type": "profile_picture", "group_id": None,
                    "body": {"image_b64": base64.b64encode(image_bytes).decode(), "mime": mime}}
        for recipient in self._known_recipients():
            await self._send_envelope_to(recipient, envelope)

    def avatar(self, username: str) -> bytes | None:
        return self.storage.get_avatar(username)

    # -- receiving ------------------------------------------------------------------------

    def _decrypt_envelope(self, incoming: IncomingMessage) -> dict:
        sender = incoming.sender
        payload = incoming.payload
        session = self.storage.load_session(sender)

        if payload.get("x3dh_header") is not None and session is None:
            header = x3dh.InitialMessageHeader.from_json(payload["x3dh_header"])
            result = x3dh.respond(self.keystore, header)
            session = RatchetState.init_as_responder(
                shared_secret=result.shared_secret,
                associated_data=result.associated_data,
                own_ratchet_priv=self.keystore.signed_prekey.private_key,
            )
            # the one-time prekey consumed by x3dh.respond() was removed from
            # the in-memory keystore -- persist that so it can't be reused.
            self.storage.save_keystore(self.keystore)

        if session is None:
            raise ValueError(f"received a message from {sender} with no session and no X3DH init header")

        ratchet_message = RatchetMessage.from_json(payload["ratchet_message"])
        envelope = _envelope_from_bytes(session.decrypt(ratchet_message))
        self.storage.save_session(sender, session)
        return envelope

    def handle_incoming(self, incoming: IncomingMessage) -> IncomingEvent:
        sender = incoming.sender
        ts = incoming.created_at
        envelope = self._decrypt_envelope(incoming)
        envelope_type = envelope["type"]
        group_id = envelope.get("group_id")
        conversation_id = group_conversation_id(group_id) if group_id else sender

        handler = self._ENVELOPE_HANDLERS.get(envelope_type)
        if handler is None:
            raise ValueError(f"unknown envelope type: {envelope_type}")

        if envelope_type not in ("group_invite", "profile_picture") and group_id and not self.storage.group_name(group_id):
            # message for a group we don't know about yet (e.g. local state lost) --
            # degrade gracefully instead of dropping the message.
            self.storage.create_group(group_id, "Unknown group", [sender, self.username])

        return handler(self, sender, conversation_id, group_id, envelope, ts)

    def _on_group_invite(self, sender, _conversation_id, _group_id, envelope, ts) -> IncomingEvent:
        body = envelope["body"]
        self.storage.create_group(body["group_id"], body["name"], body["members"])
        return IncomingEvent(kind="group_invite", conversation_id=group_conversation_id(body["group_id"]),
                              sender=sender, timestamp=ts, body=body)

    def _on_profile_picture(self, sender, _conversation_id, _group_id, envelope, ts) -> IncomingEvent:
        image_bytes = base64.b64decode(envelope["body"]["image_b64"])
        self.storage.set_avatar(sender, image_bytes)
        return IncomingEvent(kind="avatar_update", conversation_id=sender, sender=sender, timestamp=ts,
                              body=envelope["body"])

    def _on_delete(self, sender, conversation_id, _group_id, envelope, ts) -> IncomingEvent:
        target_id = envelope["body"]["target_id"]
        self.storage.mark_deleted(conversation_id, target_id)
        return IncomingEvent(kind="delete", conversation_id=conversation_id, sender=sender,
                              timestamp=ts, message_id=target_id, body=envelope["body"])

    def _on_group_member_added(self, sender, conversation_id, group_id, envelope, ts) -> IncomingEvent:
        added_username = envelope["body"]["added_username"]
        self.storage.add_group_member(group_id, added_username)
        self.storage.add_message(
            conversation_id, None, "received", "system", str(uuid.uuid4()),
            {"content": f"{added_username} was added"}, timestamp=ts,
        )
        return IncomingEvent(kind="group_member_added", conversation_id=conversation_id, sender=sender,
                              timestamp=ts, body=envelope["body"])

    def _on_group_member_removed(self, sender, conversation_id, group_id, envelope, ts) -> IncomingEvent:
        removed_username = envelope["body"]["removed_username"]
        self.storage.remove_group_member(group_id, removed_username)
        note = ("You were removed from the group" if removed_username == self.username
                else f"{removed_username} was removed")
        self.storage.add_message(
            conversation_id, None, "received", "system", str(uuid.uuid4()),
            {"content": note}, timestamp=ts,
        )
        return IncomingEvent(kind="group_member_removed", conversation_id=conversation_id, sender=sender,
                              timestamp=ts, body=envelope["body"])

    def _on_content_message(self, sender, conversation_id, _group_id, envelope, ts) -> IncomingEvent:
        envelope_type = envelope["type"]
        if not is_group_conversation(conversation_id):
            self.storage.add_contact(sender)
        self.storage.add_message(
            conversation_id, sender, "received", envelope_type, envelope["id"], envelope["body"],
            timestamp=ts,
        )
        return IncomingEvent(kind="message", conversation_id=conversation_id, sender=sender, timestamp=ts,
                              message_id=envelope["id"], message_kind=envelope_type, body=envelope["body"])

    _ENVELOPE_HANDLERS = {
        "group_invite": _on_group_invite,
        "profile_picture": _on_profile_picture,
        "delete": _on_delete,
        "group_member_added": _on_group_member_added,
        "group_member_removed": _on_group_member_removed,
        "text": _on_content_message,
        "attachment": _on_content_message,
    }

    # -- read helpers -----------------------------------------------------------------------

    def history(self, conversation_id: str) -> list[StoredMessage]:
        return self.storage.get_messages(conversation_id)

    def conversations(self) -> list[ConversationSummary]:
        return self.storage.list_conversations()

    def contacts(self) -> list[str]:
        return self.storage.list_contacts()

    def set_pinned(self, conversation_id: str, pinned: bool) -> None:
        # purely a local organizational preference -- not synced to the other side(s).
        self.storage.set_pinned(conversation_id, pinned)


def _envelope_bytes(envelope: dict) -> bytes:
    return json.dumps(envelope).encode("utf-8")


def _envelope_from_bytes(data: bytes) -> dict:
    return json.loads(data.decode("utf-8"))

File: frontend/src/engine/sessionManager.ts
// Ties crypto (X3DH + Double Ratchet) to local storage and the network --
// TypeScript port of client/session_manager.py.
//
// Every ratchet-encrypted plaintext is a small JSON "envelope" (see
// buildEnvelope call sites) rather than a raw string, which is what makes
// deletes, group messages and attachments possible on top of the same
// pairwise Double Ratchet sessions built for 1:1 chat. Group conversations
// are pure pairwise fan-out: one envelope (one id) gets encrypted separately
// to every other member's existing/newly-established 1:1 session -- there
// is no group-level cryptographic key.

import { utf8Decode, utf8Encode } from './crypto/encoding'
import { type KeyStore, topUpOneTimePrekeys } from './crypto/identity'
import { initAsInitiator, initAsResponder, ratchetDecrypt, ratchetEncrypt } from './crypto/ratchet'
import { initiate, respond } from './crypto/x3dh'
import { type IncomingMessage, type NetworkClient } from './network'
import {
  type ConversationSummary,
  type LocalStorage,
  type StoredMessage,
  groupConversationId,
  isGroupConversation,
} from './storage'

const ONE_TIME_PREKEY_TARGET = 20
const ATTACHMENT_MAX_BYTES = 10 * 1024 * 1024
const AVATAR_MAX_BYTES = 2 * 1024 * 1024

export type IncomingEventKind =
  | 'message'
  | 'delete'
  | 'group_invite'
  | 'group_member_added'
  | 'group_member_removed'
  | 'avatar_update'

export interface IncomingEvent {
  kind: IncomingEventKind
  conversationId: string
  sender: string
  timestamp: number
  messageId?: string | null
  messageKind?: string | null // "text" | "attachment" | "system", when kind === "message"
  body: Record<string, unknown>
}

function base64FromBytes(bytes: Uint8Array): string {
  let binary = ''
  for (let i = 0; i < bytes.length; i++) binary += String.fromCharCode(bytes[i])
  return btoa(binary)
}

function bytesFromBase64(b64: string): Uint8Array {
  const binary = atob(b64)
  const bytes = new Uint8Array(binary.length)
  for (let i = 0; i < binary.length; i++) bytes[i] = binary.charCodeAt(i)
  return bytes
}

function envelopeBytes(envelope: Record<string, unknown>): Uint8Array {
  return utf8Encode(JSON.stringify(envelope))
}

function envelopeFromBytes(data: Uint8Array): Record<string, unknown> {
  return JSON.parse(utf8Decode(data))
}

type EnvelopeHandler = (
  manager: SessionManager,
  sender: string,
  conversationId: string,
  groupId: string | null,
  envelope: Record<string, unknown>,
  timestamp: number,
) => Promise<IncomingEvent>

export class SessionManager {
  readonly keystore: KeyStore
  readonly storage: LocalStorage
  readonly network: NetworkClient

  constructor(keystore: KeyStore, storage: LocalStorage, network: NetworkClient) {
    this.keystore = keystore
    this.storage = storage
    this.network = network
  }

  get username(): string {
    return this.keystore.username
  }

  /** Upload identity/signed-prekey/one-time-prekeys; top up if running low. */
  async publishKeys(): Promise<void> {
    if (topUpOneTimePrekeys(this.keystore, ONE_TIME_PREKEY_TARGET).length > 0) {
      await this.storage.saveKeystore(this.keystore)
    }
    await this.network.uploadKeys(this.keystore)
  }

  // -- recipient resolution --------------------------------------------------------

  private async recipientsFor(conversationId: string): Promise<string[]> {
    if (isGroupConversation(conversationId)) {
      const groupId = conversationId.split(':').slice(1).join(':')
      const members = await this.storage.groupMembers(groupId)
      return members.filter((m) => m !== this.username)
    }
    return [conversationId]
  }

  // -- low-level: encrypt+send one envelope to one recipient -----------------------

  private async sendEnvelopeTo(recipientUsername: string, envelope: Record<string, unknown>): Promise<void> {
    let session = await this.storage.loadSession(recipientUsername)
    let x3dhHeaderJson: Record<string, unknown> | null = null

    if (session === null) {
      const bundle = await this.network.fetchBundle(recipientUsername)
      if (bundle === null) {
        throw new Error(`no such user or no published keys: ${recipientUsername}`)
      }
      const { result, header } = initiate(this.keystore, bundle)
      session = initAsInitiator(result.sharedSecret, result.associatedData, bundle.signedPrekeyPub)
      x3dhHeaderJson = {
        initiator_identity_pub_dh: base64FromBytes(header.initiatorIdentityPubDh),
        initiator_ephemeral_pub: base64FromBytes(header.initiatorEphemeralPub),
        signed_prekey_id: header.signedPrekeyId,
        one_time_prekey_id: header.oneTimePrekeyId,
      }
    }

    const ratchetMessage = ratchetEncrypt(session, envelopeBytes(envelope))
    const payload = {
      x3dh_header: x3dhHeaderJson,
      ratchet_message: {
        header: {
          dh_pub: base64FromBytes(ratchetMessage.header.dhPub),
          pn: ratchetMessage.header.pn,
          n: ratchetMessage.header.n,
        },
        ciphertext: base64FromBytes(ratchetMessage.ciphertext),
      },
    }
    await this.network.sendMessage(recipientUsername, payload)
    await this.storage.saveSession(recipientUsername, session)
  }

  private async fanOut(conversationId: string, envelope: Record<string, unknown>): Promise<void> {
    for (const recipient of await this.recipientsFor(conversationId)) {
      await this.sendEnvelopeTo(recipient, envelope)
    }
  }

  private async requireGroupMembership(conversationId: string): Promise<void> {
    if (isGroupConversation(conversationId)) {
      const groupId = conversationId.split(':').slice(1).join(':')
      const members = await this.storage.groupMembers(groupId)
      if (!members.includes(this.username)) {
        throw new Error('you are no longer a member of this group')
      }
    }
  }

  private groupIdOf(conversationId: string): string | null {
    return isGroupConversation(conversationId) ? conversationId.split(':').slice(1).join(':') : null
  }

  // -- sending: text / attachment ---------------------------------------------------

  async sendText(conversationId: string, text: string): Promise<string> {
    await this.requireGroupMembership(conversationId)
    const messageId = crypto.randomUUID()
    const groupId = this.groupIdOf(conversationId)
    const envelope = { id: messageId, type: 'text', group_id: groupId, body: { content: text } }
    await this.fanOut(conversationId, envelope)

    if (!isGroupConversation(conversationId)) {
      await this.storage.addContact(conversationId)
    }
    await this.storage.addMessage(conversationId, this.username, 'sent', 'text', messageId, envelope.body, Date.now() / 1000)
    return messageId
  }

  async sendAttachment(conversationId: string, filename: string, mime: string, data: Uint8Array): Promise<string> {
    await this.requireGroupMembership(conversationId)
    if (data.length > ATTACHMENT_MAX_BYTES) {
      throw new Error(`attachment too large (${data.length} bytes, max ${ATTACHMENT_MAX_BYTES})`)
    }
    const messageId = crypto.randomUUID()
    const groupId = this.groupIdOf(conversationId)
    const body = { filename, mime, size: data.length, data_b64: base64FromBytes(data) }
    const envelope = { id: messageId, type: 'attachment', group_id: groupId, body }
    await this.fanOut(conversationId, envelope)

    if (!isGroupConversation(conversationId)) {
      await this.storage.addContact(conversationId)
    }
    await this.storage.addMessage(conversationId, this.username, 'sent', 'attachment', messageId, body, Date.now() / 1000)
    return messageId
  }

  // -- delete-for-everyone -----------------------------------------------------------

  async deleteMessage(conversationId: string, messageId: string): Promise<void> {
    await this.storage.markDeleted(conversationId, messageId)
    const groupId = this.groupIdOf(conversationId)
    const envelope = { id: crypto.randomUUID(), type: 'delete', group_id: groupId, body: { target_id: messageId } }
    await this.fanOut(conversationId, envelope)
  }

  // -- groups ---------------------------------------------------------------------------

  async createGroup(name: string, members: string[]): Promise<string> {
    const groupId = crypto.randomUUID()
    const allMembers = Array.from(new Set([...members, this.username])).sort()
    await this.storage.createGroup(groupId, name, allMembers)
    const conversationId = groupConversationId(groupId)

    const envelope = { id: crypto.randomUUID(), type: 'group_invite', group_id: groupId, body: { group_id: groupId, name, members: allMembers } }
    for (const member of allMembers) {
      if (member !== this.username) await this.sendEnvelopeTo(member, envelope)
    }

    await this.storage.addMessage(conversationId, null, 'sent', 'system', crypto.randomUUID(), { content: 'You created this group' }, Date.now() / 1000)
    return groupId
  }

  async groups(): Promise<[string, string][]> {
    return this.storage.listGroups()
  }

  async addGroupMember(groupId: string, username: string): Promise<void> {
    const conversationId = groupConversationId(groupId)
    const currentMembers = await this.storage.groupMembers(groupId)
    if (currentMembers.includes(username)) return
    await this.storage.addGroupMember(groupId, username)
    const newRoster = Array.from(new Set([...currentMembers, username])).sort()
    const name = (await this.storage.groupName(groupId)) ?? 'Unnamed group'

    const inviteEnvelope = { id: crypto.randomUUID(), type: 'group_invite', group_id: groupId, body: { group_id: groupId, name, members: newRoster } }
    await this.sendEnvelopeTo(username, inviteEnvelope)

    const addedEnvelope = { id: crypto.randomUUID(), type: 'group_member_added', group_id: groupId, body: { group_id: groupId, added_username: username } }
    for (const member of currentMembers) {
      if (member !== this.username) await this.sendEnvelopeTo(member, addedEnvelope)
    }

    await this.storage.addMessage(conversationId, null, 'sent', 'system', crypto.randomUUID(), { content: `You added ${username}` }, Date.now() / 1000)
  }

  async removeGroupMember(groupId: string, username: string): Promise<void> {
    const conversationId = groupConversationId(groupId)
    const oldRoster = await this.storage.groupMembers(groupId)
    if (!oldRoster.includes(username)) return
    await this.storage.removeGroupMember(groupId, username)

    const removedEnvelope = { id: crypto.randomUUID(), type: 'group_member_removed', group_id: groupId, body: { group_id: groupId, removed_username: username } }
    for (const member of oldRoster) {
      if (member !== this.username) await this.sendEnvelopeTo(member, removedEnvelope)
    }

    const note = username === this.username ? 'You left this group' : `You removed ${username}`
    await this.storage.addMessage(conversationId, null, 'sent', 'system', crypto.randomUUID(), { content: note }, Date.now() / 1000)
  }

  // -- profile pictures -------------------------------------------------------------------

  private async knownRecipients(): Promise<Set<string>> {
    const recipients = new Set(await this.storage.listContacts())
    for (const [groupId] of await this.storage.listGroups()) {
      for (const member of await this.storage.groupMembers(groupId)) recipients.add(member)
    }
    recipients.delete(this.username)
    return recipients
  }

  async setProfilePicture(imageBytes: Uint8Array, mime: string): Promise<void> {
    if (imageBytes.length > AVATAR_MAX_BYTES) {
      throw new Error(`image too large (${imageBytes.length} bytes, max ${AVATAR_MAX_BYTES})`)
    }
    await this.storage.setAvatar(this.username, imageBytes)
    const envelope = { id: crypto.randomUUID(), type: 'profile_picture', group_id: null, body: { image_b64: base64FromBytes(imageBytes), mime } }
    for (const recipient of await this.knownRecipients()) {
      await this.sendEnvelopeTo(recipient, envelope)
    }
  }

  async avatar(username: string): Promise<Uint8Array | null> {
    return this.storage.getAvatar(username)
  }

  // -- receiving ------------------------------------------------------------------------

  private async decryptEnvelope(incoming: IncomingMessage): Promise<Record<string, unknown>> {
    const sender = incoming.sender
    const payload = incoming.payload
    let session = await this.storage.loadSession(sender)

    if (payload.x3dh_header !== null && payload.x3dh_header !== undefined && session === null) {
      const headerJson = payload.x3dh_header as Record<string, unknown>
      const header = {
        initiatorIdentityPubDh: bytesFromBase64(headerJson.initiator_identity_pub_dh as string),
        initiatorEphemeralPub: bytesFromBase64(headerJson.initiator_ephemeral_pub as string),
        signedPrekeyId: headerJson.signed_prekey_id as string,
        oneTimePrekeyId: (headerJson.one_time_prekey_id as string | null) ?? null,
      }
      const result = respond(this.keystore, header)
      session = initAsResponder(result.sharedSecret, result.associatedData, this.keystore.signedPrekey.privateKey)
      // the one-time prekey consumed by respond() was removed from the
      // in-memory keystore -- persist that so it can't be reused.
      await this.storage.saveKeystore(this.keystore)
    }

    if (session === null) {
      throw new Error(`received a message from ${sender} with no session and no X3DH init header`)
    }

    const ratchetMessageJson = payload.ratchet_message as Record<string, unknown>
    const headerJson = ratchetMessageJson.header as Record<string, unknown>
    const ratchetMessage = {
      header: { dhPub: bytesFromBase64(headerJson.dh_pub as string), pn: headerJson.pn as number, n: headerJson.n as number },
      ciphertext: bytesFromBase64(ratchetMessageJson.ciphertext as string),
    }
    const envelope = envelopeFromBytes(ratchetDecrypt(session, ratchetMessage))
    await this.storage.saveSession(sender, session)
    return envelope
  }

  async handleIncoming(incoming: IncomingMessage): Promise<IncomingEvent> {
    const sender = incoming.sender
    const ts = incoming.createdAt
    const envelope = await this.decryptEnvelope(incoming)
    const envelopeType = envelope.type as string
    const groupId = (envelope.group_id as string | null) ?? null
    const conversationId = groupId ? groupConversationId(groupId) : sender

    const handler = ENVELOPE_HANDLERS[envelopeType]
    if (handler === undefined) {
      throw new Error(`unknown envelope type: ${envelopeType}`)
    }

    if (envelopeType !== 'group_invite' && envelopeType !== 'profile_picture' && groupId && !(await this.storage.groupName(groupId))) {
      // message for a group we don't know about yet (e.g. local state lost) --
      // degrade gracefully instead of dropping the message.
      await this.storage.createGroup(groupId, 'Unknown group', [sender, this.username])
    }

    return await handler(this, sender, conversationId, groupId, envelope, ts)
  }

  // -- read helpers -----------------------------------------------------------------------

  async history(conversationId: string): Promise<StoredMessage[]> {
    return this.storage.getMessages(conversationId)
  }

  async conversations(): Promise<ConversationSummary[]> {
    return this.storage.listConversations()
  }

  async contacts(): Promise<string[]> {
    return this.storage.listContacts()
  }

  async setPinned(conversationId: string, pinned: boolean): Promise<void> {
    // purely a local organizational preference -- not synced to the other side(s).
    await this.storage.setPinned(conversationId, pinned)
  }
}

async function onGroupInvite(manager: SessionManager, sender: string, _conversationId: string, _groupId: string | null, envelope: Record<string, unknown>, ts: number): Promise<IncomingEvent> {
  const body = envelope.body as Record<string, unknown>
  await manager.storage.createGroup(body.group_id as string, body.name as string, body.members as string[])
  return { kind: 'group_invite', conversationId: groupConversationId(body.group_id as string), sender, timestamp: ts, body }
}

async function onProfilePicture(manager: SessionManager, sender: string, _conversationId: string, _groupId: string | null, envelope: Record<string, unknown>, ts: number): Promise<IncomingEvent> {
  const body = envelope.body as Record<string, unknown>
  const imageBytes = bytesFromBase64(body.image_b64 as string)
  await manager.storage.setAvatar(sender, imageBytes)
  return { kind: 'avatar_update', conversationId: sender, sender, timestamp: ts, body }
}

async function onDelete(manager: SessionManager, sender: string, conversationId: string, _groupId: string | null, envelope: Record<string, unknown>, ts: number): Promise<IncomingEvent> {
  const body = envelope.body as Record<string, unknown>
  const targetId = body.target_id as string
  await manager.storage.markDeleted(conversationId, targetId)
  return { kind: 'delete', conversationId, sender, timestamp: ts, messageId: targetId, body }
}

async function onGroupMemberAdded(manager: SessionManager, sender: string, conversationId: string, groupId: string | null, envelope: Record<string, unknown>, ts: number): Promise<IncomingEvent> {
  const body = envelope.body as Record<string, unknown>
  const addedUsername = body.added_username as string
  await manager.storage.addGroupMember(groupId as string, addedUsername)
  await manager.storage.addMessage(conversationId, null, 'received', 'system', crypto.randomUUID(), { content: `${addedUsername} was added` }, ts)
  return { kind: 'group_member_added', conversationId, sender, timestamp: ts, body }
}

async function onGroupMemberRemoved(manager: SessionManager, sender: string, conversationId: string, groupId: string | null, envelope: Record<string, unknown>, ts: number): Promise<IncomingEvent> {
  const body = envelope.body as Record<string, unknown>
  const removedUsername = body.removed_username as string
  await manager.storage.removeGroupMember(groupId as string, removedUsername)
  const note = removedUsername === manager.username ? 'You were removed from the group' : `${removedUsername} was removed`
  await manager.storage.addMessage(conversationId, null, 'received', 'system', crypto.randomUUID(), { content: note }, ts)
  return { kind: 'group_member_removed', conversationId, sender, timestamp: ts, body }
}

async function onContentMessage(manager: SessionManager, sender: string, conversationId: string, _groupId: string | null, envelope: Record<string, unknown>, ts: number): Promise<IncomingEvent> {
  const envelopeType = envelope.type as string
  if (!isGroupConversation(conversationId)) {
    await manager.storage.addContact(sender)
  }
  const body = envelope.body as Record<string, unknown>
  await manager.storage.addMessage(conversationId, sender, 'received', envelopeType, envelope.id as string, body, ts)
  return { kind: 'message', conversationId, sender, timestamp: ts, messageId: envelope.id as string, messageKind: envelopeType, body }
}

const ENVELOPE_HANDLERS: Record<string, EnvelopeHandler> = {
  group_invite: onGroupInvite,
  profile_picture: onProfilePicture,
  delete: onDelete,
  group_member_added: onGroupMemberAdded,
  group_member_removed: onGroupMemberRemoved,
  text: onContentMessage,
  attachment: onContentMessage,
}
